rpcs3/Emu/Cell/lv2/sys_rsxaudio.cpp

```diff
@@ -1,39 +1,2203 @@
+			ringbuf.set_buf_size(frame_len_bytes);
+			tmp_buf = std::make_unique<u8[]>(frame_len_bytes);
+		}
+		else
+		{
+			const f64 frame_len = std::max<f64>(buffering_len, cb_frame_len) + _10ms;
+			ringbuf.set_buf_size(static_cast<u64>(std::round(frame_len * bytes_per_sec)));
+			tmp_buf = nullptr;
+		}
+	}
+
+	callback_cfg.atomic_op([&](callback_config &val)
+	{
+		val.callback_active = false; // Backend may take some time to activate. This prevents overflows on input side.
+
+		if (!val.cfg_changed)
+		{
+			val.freq = static_cast<u32>(port_cfg.freq);
+			val.input_ch_cnt = static_cast<u32>(port_cfg.ch_cnt);
+			val.output_ch_cnt = static_cast<u32>(ch_cnt);
+			val.convert_to_s16 = emu_cfg.convert_to_s16;
+			val.avport_idx = emu_cfg.avport;
+			val.ready = true;
+		}
+	});
+}
+
+void rsxaudio_backend_thread::backend_start()
+{
+	ensure(backend != nullptr);
+
+	if (use_aux_ringbuf)
+	{
+		resampler.set_tempo(RESAMPLER_MAX_FREQ_VAL);
+		resampler.flush();
+		aux_ringbuf.reader_flush();
+	}
+
+	ringbuf.reader_flush();
+	backend->Play();
+}
+
+void rsxaudio_backend_thread::backend_stop()
+{
+	if (backend == nullptr)
+	{
+		return;
+	}
+
+	backend->Pause();
+	callback_cfg.atomic_op([&](callback_config &val)
+	{
+		val.callback_active = false;
+	});
+}
+
+bool rsxaudio_backend_thread::backend_playing()
+{
+	if (backend == nullptr)
+	{
+		return false;
+	}
+
+	return backend->IsPlaying();
+}
+
+u32 rsxaudio_backend_thread::write_data_callback(u32 bytes, void *buf)
+{
+	const callback_config cb_cfg = callback_cfg.atomic_op([&](callback_config &val)
+	{
+		val.callback_active = true;
+		return val;
+	});
+
+	const std::bitset<SYS_RSXAUDIO_AVPORT_CNT> mute_state{cb_cfg.mute_state};
+
+	if (cb_cfg.ready && !mute_state[static_cast<u8>(cb_cfg.avport_idx)] && Emu.IsRunning())
+	{
+		const u32 bytes_from_rb = cb_cfg.convert_to_s16 ? bytes / static_cast<u32>(AudioSampleSize::S16) * static_cast<u32>(AudioSampleSize::FLOAT) : bytes;
+		const u32 byte_cnt = static_cast<u32>(ringbuf.pop(buf, bytes_from_rb, true));
+		const u32 sample_cnt = byte_cnt / static_cast<u32>(AudioSampleSize::FLOAT);
+
+		// Buffer is in weird state - drop acquired data
+		if (sample_cnt == 0 || sample_cnt % cb_cfg.input_ch_cnt != 0)
+		{
+			memset(buf, 0, bytes);
+			return bytes;
+		}
+
+		if (cb_cfg.input_ch_cnt > cb_cfg.output_ch_cnt)
+		{
+			if (cb_cfg.input_ch_cnt == static_cast<u32>(AudioChannelCnt::SURROUND_7_1))
+			{
+				if (cb_cfg.output_ch_cnt == static_cast<u32>(AudioChannelCnt::SURROUND_5_1))
+				{
+					AudioBackend::downmix<AudioChannelCnt::SURROUND_7_1, AudioChannelCnt::SURROUND_5_1>(sample_cnt, static_cast<f32*>(buf), static_cast<f32*>(buf));
+				}
+				else if (cb_cfg.output_ch_cnt == static_cast<u32>(AudioChannelCnt::STEREO))
+				{
+					AudioBackend::downmix<AudioChannelCnt::SURROUND_7_1, AudioChannelCnt::STEREO>(sample_cnt, static_cast<f32*>(buf), static_cast<f32*>(buf));
+				}
+				else
+				{
+					ensure(false);
```

Comment:
always use fmt::throw_exception when you do ensure(false).
ensure false doesn't tell us any information. apart from the line maybe